From 006743f0dfd284b5ae1d16869533f845e1b2394c Mon Sep 17 00:00:00 2001
From: appkins <13772502+appkins@users.noreply.github.com>
Date: Mon, 16 Mar 2026 13:02:41 -0500
Subject: [PATCH 01/11] arm: bcm283x: add BCM2712 SoC support

Add Kconfig entry for BCM2712 SoC with driver selections for RP1
peripherals, ACPI table definitions, and board type constants
needed for Raspberry Pi 5 support.

Link: https://github.com/sidero-community/u-boot
Signed-off-by: appkins <13772502+appkins@users.noreply.github.com>
[Andrew Regenscheid: rebased onto v2026.07; dropped the init.c memory-map
 hunks (superseded by upstream b20ce94b "ARM: bcm283x: Add bcm2712 PCIe
 memory window"), the bcm2835.c bcm2712-hdmi0 compatible (already upstream),
 and the bcm2711.h PCIe register additions (only consumed by the brcmstb
 driver rewrite that upstream has since superseded).]
Signed-off-by: Andrew Regenscheid <andrew@regenscheid.net>
---
 arch/arm/mach-bcm283x/Kconfig      | 16 +++++++++++++++
 include/broadcom/bcm_board_types.h | 32 ++++++++++++++++++++++++++++++
 2 files changed, 48 insertions(+)
 create mode 100644 include/broadcom/bcm_board_types.h

diff --git a/arch/arm/mach-bcm283x/Kconfig b/arch/arm/mach-bcm283x/Kconfig
index d9303e8c..861d0d9a 100644
--- a/arch/arm/mach-bcm283x/Kconfig
+++ b/arch/arm/mach-bcm283x/Kconfig
@@ -56,6 +56,22 @@ config BCM2711_64B
 	select BLOBLIST_ALLOC if GENERATE_ACPI_TABLE
 	select BLOBLIST_TABLES if GENERATE_ACPI_TABLE
 
+config BCM2712
+	bool "Broadcom BCM2712 SoC support"
+	depends on ARCH_BCM283X
+	select DRIVER_GICV2
+	select ARM64
+	select CPU_ARMV8
+	select PINCTRL_BCM2712 if PINCTRL
+	select RP1_GPIO
+	select CLK_RP1 if CLK
+	select ARMV8_MULTIENTRY if GENERATE_ACPI_TABLE
+	select BLOBLIST if GENERATE_ACPI_TABLE
+	select BLOBLIST_ALLOC if GENERATE_ACPI_TABLE
+	select BLOBLIST_TABLES if GENERATE_ACPI_TABLE
+	imply OF_UPSTREAM
+	imply BOARD_TYPES
+
 menu "Broadcom BCM283X family"
 	depends on ARCH_BCM283X
 choice
diff --git a/include/broadcom/bcm_board_types.h b/include/broadcom/bcm_board_types.h
new file mode 100644
index 00000000..3508b40f
--- /dev/null
+++ b/include/broadcom/bcm_board_types.h
@@ -0,0 +1,32 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * This header defines BCM board types.
+ *
+ * Copyright (c) 2024 EPAM Systems
+ *
+ */
+#ifndef _BCM_BOARD_TYPES_H_
+#define _BCM_BOARD_TYPES_H_
+
+/*
+ * TODO: This enumerator defines the main BCM types.
+ * These types are defined by the BCM firmware and requested
+ * via mbox. Currently only one board type was added.
+ * This is needed to determine whether current board is bcm2712
+ * or other to control the behaviour of the following drivers:
+ * macb and pcie_brcmstb. New board enumerators can be added here
+ * if needed.
+ * See https://www.raspberrypi.com/documentation/computers/raspberry-pi.html#raspberry-pi-revision-codes
+ */
+typedef enum {
+	BCM2712_RPI_5_B_NEW = 0x17,
+	BCM2712_CM5 = 0x18,
+	BCM2712_RPI_500 = 0x19,
+	BCM2712_CM5_LITE = 0x1A,
+} bcm_board_type_t;
+
+/* Range constants for Raspberry Pi 5 family boards */
+#define RPI_BOARD_TYPE_RPI5_FAMILY_MIN	BCM2712_RPI_5_B_NEW
+#define RPI_BOARD_TYPE_RPI5_FAMILY_MAX	BCM2712_CM5_LITE
+
+#endif /* _BCM_BOARD_TYPES_H_ */
